server/src/storage/datastore.ts

```diff
@@ -64,35 +63,68 @@ const getAll = async (kind: string, filters?: Filter[]) => {
 };
 
 /**
- * Unpacks id of modified object from MutationResult object.
+ * A Datastore wrapper that inserts a particular entity with the specified Kind if another
+ * entity with the same value for the specified unique property does not already exist.
+ * An error is thrown if an entity with the same unique property value already exists.
+ * @param kind The Kind of the Entity
+ * @param entity The Entity to be added
+ * @param uniqueProperty The property that should be unique for the specified kind
  */
-const getIdFromMutationResult = (
-  mutationResult: google.datastore.v1.IMutationResult
-): ResponseId | undefined => mutationResult?.key?.path?.[0]?.id;
+const insertUniqueEntity = async (
+  kind: string,
+  entity: Entity,
+  uniqueProperty: Filter
+): Promise<Entity | undefined> => {
+  const transaction = datastore.transaction();
+  const query = transaction
+    .createQuery(kind)
+    .filter(uniqueProperty.property, uniqueProperty.value);
 
-/**
- * Unpacks ids of modified objects from CommitResponse object received from Datastore.
- */
-const getIdsFromCommitResponse = (
-  res: google.datastore.v1.ICommitResponse
-): (ResponseId | undefined)[] | undefined =>
-  res.mutationResults?.map(getIdFromMutationResult);
+  try {
+    await transaction.run();
+    const [queryRes] = await transaction.runQuery(query);
+    if (queryRes && queryRes.length > 0) {
+      // An entity with the unqiue property already exists
+      await transaction.rollback();
+      throw new Error(
```

Comment:
this error should be caught outside, so you rollback twice?

Reply:
@ifhuang Oh okay I will remove this rollback and just leave the throwing of error here! Previously I wanted this to return something instead of throwing an error so I added the rollback call. But now that I have decided to throw an error instead, it is redundant as you have pointed out!